README.md
# Pincelate

By [Allison Parrish](https://www.decontextualize.com/)

Pincelate is a machine learning model for spelling English words and sounding
them out, plus a Python module that makes it super simple to do fun and useful
things with the model.

A quick example:

    >>> from pincelate import Pincelate
    >>> pin = Pincelate()
    >>> pin.soundout("pincelate")
    ['P', 'IH1', 'N', 'S', 'AH0', 'L', 'EY1', 'T']
    >>> pin.spell(["HH", "EH", "L", "OW"])
    'hello'

Please see [the documentation](https://pincelate.readthedocs.io/en/latest/) for
more information!



source/changelog.rst
Changelog
=========

0.0.1 (2019-11-05)
------------------

* Initial release



tests/test_seq2seq.py
import unittest
import pincelate
import pincelate.cmudictdata


class TestSeq2Seq(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        print("loading data")
        cls.data = pincelate.cmudictdata.load()
        print("done")
        cls.tts = pincelate.cmudictdata.tts(cls.data)

    def test_train(self):
        from pincelate.seq2seq import Seq2Seq
        testmodel = Seq2Seq(
                src_vocab=self.data['orthography_vocab'],
                target_vocab=self.data['phoneme_feature_vocab'],
                activation='sigmoid',
                loss='binary_crossentropy',
                hps=dict(
                    optimizer='adam',
                    lr=0.001,
                    bidi=True,
                    enc_rnn_units=4,
                    dec_rnn_units=4,
                    enc_rnn_dropout=0.2,
                    dec_rnn_dropout=0.2,
                    vae=False
                )
        )
        testmodel.make_models()
        testmodel.compile_models()
        print(testmodel.training_model.summary())
        history = testmodel.fit(
                [self.tts['orth_data_train'][:1000],
                    self.tts['phon_data_train'][:1000]],
                self.tts['phon_target_train'][:1000],
                batch_size=256,
                epochs=1,
                validation_data=(
                    [self.tts['orth_data_valid'][:100],
                        self.tts['phon_data_valid'][:100]],
                    self.tts['phon_target_valid'][:100]))
        testmodel.save("tests/testmodel")
        loaded = Seq2Seq.load("tests/testmodel")

        # ensure that histories are identical
        self.assertEqual(loaded.training_history, history.history)

        # all fields should be identical
        for field in Seq2Seq.SERIALIZE_FIELDS:
            self.assertEqual(
                    getattr(loaded, field),
                    getattr(testmodel, field))


if __name__ == '__main__':
    unittest.main()


tests/test_pincelate.py
import unittest
import numpy as np
import os
from os.path import join as opj

from pincelate import Pincelate

TEST_DIR = os.path.dirname(os.path.abspath(__file__))


class TestPincelate(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        cls.pin = Pincelate()

    def test_load_custom_models(self):
        from pincelate import Pincelate
        pin = Pincelate(
                ('pincelate/models/orth-phon-enc256-dec256',
                 'pincelate/models/phon-orth-enc256-dec256'))
        self.assertEqual(pin.soundout("hello"), ['HH', 'EH1', 'L', 'OW0'])

    def test_soundout(self):
        self.assertEqual(self.pin.soundout("hello"), ['HH', 'EH1', 'L', 'OW0'])

    def test_spell(self):
        self.assertEqual(self.pin.spell(['HH', 'EH1', 'L', 'OW0']), 'hello')

    def test_phonemefeatures(self):
        with open(opj(TEST_DIR, "allison-phoneme-features.npy"), "rb") as fh:
            arr = np.load(fh)
            self.assertTrue(
                    np.all(self.pin.phonemefeatures("allison") == arr))
            self.assertIn(self.pin.spellfeatures(arr), ("allison", "alison"))

    def test_phonemestate(self):
        with open(opj(TEST_DIR, "allison-phoneme-state.npy"), "rb") as fh:
            arr = np.load(fh)[0]
            self.assertTrue(np.all(self.pin.phonemestate("allison") == arr))
            self.assertIn(self.pin.spellstate(arr), ("allison", "alison"))

    def test_manipulate(self):
        res = self.pin.manipulate("illicit", letters={'i': 10, 'y': -2},
                                  temperature=0.05)
        self.assertGreater(res.count('y'), res.count('i'))
        res = self.pin.manipulate("ticktock", features={'nas': -5, 'vcd': -5},
                                  temperature=0.05)
        self.assertGreaterEqual(res.count('d'), res.count('t'))
        self.assertGreater(res.count('g'), res.count('k'))

    def test_vectorizefeatures(self):
        bee = self.pin.vectorizefeatures([
            ['beg'], ['blb', 'stp', 'vcd'], ['hgh', 'fnt', 'vwl'], ['end']])
        with open(opj(TEST_DIR, "bee-features.npy"), "rb") as fh:
            arr = np.load(fh)
            self.assertTrue(np.all(bee == arr))

    def test_featureidx(self):
        from pincelate.featurephone import phone_feature_map
        self.assertEqual(self.pin.featureidx('beg'),
                         self.pin.orth2phon.target_vocab_idx_map['beg'])


if __name__ == '__main__':
    unittest.main()


pincelate/seq2seq.py
from types import SimpleNamespace

from keras.models import Model
from keras.layers import Input, GRU, Dense, Masking, Concatenate, \
    Bidirectional
from keras import optimizers
import numpy as np


class Seq2Seq:
    """
    A sequence to sequence model with recurrent neural networks, plus
    some helpful functions for training, saving, loading, inference and
    decoding.
    """
    def __init__(self, src_vocab, target_vocab, activation, loss,
                 training_history=None, hps=None):
        self.src_vocab = src_vocab
        self.src_vocab_idx_map = {k: i for i, k in enumerate(src_vocab)}
        self.target_vocab = target_vocab
        self.target_vocab_idx_map = {k: i for i, k in enumerate(target_vocab)}
        self.activation = activation
        self.loss = loss
        self.training_history = training_history
        if hps is None:
            hps = {}
        if isinstance(hps, dict):
            self.hps = SimpleNamespace(**hps)
        elif isinstance(hps, SimpleNamespace):
            self.hps = hps
        else:
            raise ValueError("incompatible type for hps")

    def make_models(self):

        #
        # training models
        #

        training_encoder_inputs = Input(
            shape=(None, len(self.src_vocab)),
            name='training_encoder_inputs'
        )

        # masking to exclude zero-padding in loss/accuracy calculation
        training_encoder_mask = Masking(
                0.0,
                name='training_encoder_mask')(training_encoder_inputs)

        # two setups for encoder: bidirectional and unidirectional
        if self.hps.bidi:
            training_encoder = Bidirectional(
                    GRU(
                        self.hps.enc_rnn_units,
                        recurrent_dropout=self.hps.enc_rnn_dropout,
                        return_state=True),
                    name='training_encoder_bidi')
            # Bidirectional wrapper returns both forward and backward state
            (_,
             training_encoder_fstate,
             training_encoder_bstate) = training_encoder(training_encoder_mask)
            training_encoder_state = Concatenate(
                    name='training_encoder_bidi_state')(
                            [training_encoder_fstate, training_encoder_bstate])
        else:
            training_encoder = GRU(
                    self.hps.enc_rnn_units,
                    recurrent_dropout=self.hps.enc_rnn_dropout,
                    return_state=True,
                    name='training_encoder')
            _, training_encoder_state = training_encoder(training_encoder_mask)

        # dimensions of encoder hidden state may not match the dimensions of
        # the decoder's hidden state (and the keras implementation doesn't
        # allow us to set these separately). so we need a layer to match the
        # two.
        initial_state = Dense(self.hps.dec_rnn_units, name='initial_state')

        training_initial_state = initial_state(training_encoder_state)

        training_decoder_inputs = Input(
                shape=(None, len(self.target_vocab)),
                name='training_decoder_inputs')
        # masking to prevent zero-padding from being used in loss/accuracy
        # scores
        training_decoder_mask = Masking(
                0.0,
                name='training_decoder_mask')(training_decoder_inputs)

        training_decoder_in = training_decoder_mask

        training_decoder_rnn = GRU(
                self.hps.dec_rnn_units,
                recurrent_dropout=self.hps.dec_rnn_dropout,
                return_sequences=True,
                return_state=True,
                name='training_decoder')
        training_decoder_outputs, _ = training_decoder_rnn(
            training_decoder_in,
            initial_state=training_initial_state)
        # activation will be softmax for one-hot data, sigmoid for k-hot
        training_decoder_dense = Dense(
                len(self.target_vocab),
                activation=self.activation,
                name='training_decoder_outputs')
        training_decoder_outputs = training_decoder_dense(
                training_decoder_outputs)

        self.training_model = Model(
            [training_encoder_inputs, training_decoder_inputs],
            training_decoder_outputs)

        #
        # inference models
        #

        # encode sequence input to state for decoder
        infer_encoder_model = Model(
                training_encoder_inputs,
                training_initial_state)

        # decode hidden state and incoming sequence to target plus hidden
        # state of decoder rnn
        infer_decoder_state_input = Input(shape=(self.hps.dec_rnn_units,))
        infer_decoder_outputs, infer_decoder_state = training_decoder_rnn(
            training_decoder_inputs, initial_state=infer_decoder_state_input
        )
        infer_decoder_outputs = training_decoder_dense(infer_decoder_outputs)
        infer_decoder_model = Model(
            [training_decoder_inputs, infer_decoder_state_input],
            [infer_decoder_outputs, infer_decoder_state]
        )

        self.infer_encoder_model = infer_encoder_model
        self.infer_decoder_model = infer_decoder_model

    def compile_models(self):
        loss = self.loss
        metrics = ['accuracy']
        if self.hps.optimizer == 'rmsprop':
            optimizer = optimizers.RMSProp
        elif self.hps.optimizer == 'adam':
            optimizer = optimizers.Adam
        optimizer_params = {
            k: getattr(self.hps, k) for k in ['lr', 'decay', 'clipvalue']
            if hasattr(self.hps, k)
        }
        self.training_model.compile(
            optimizer=optimizer(**optimizer_params),
            loss=loss,
            metrics=metrics)

    SERIALIZE_FIELDS = [
            "src_vocab", "target_vocab", "activation", "loss",
            "training_history", "hps"]

    def save(self, prefix):
        import pickle
        obj = {}
        for field in self.SERIALIZE_FIELDS:
            obj[field] = getattr(self, field)
        with open(prefix + "-obj.pickle", "wb") as fh:
            pickle.dump(obj, fh)
        self.save_model_weights(prefix)

    @classmethod
    def load(cls, prefix):
        import pickle
        with open(prefix + "-obj.pickle", "rb") as fh:
            obj_data = pickle.load(fh)
        obj = cls(**obj_data)
        obj.make_models()
        obj.load_model_weights(prefix)
        return obj

    @classmethod
    def load_from_package(cls, prefix):
        from pkg_resources import resource_stream as rs
        import pickle
        with rs(__name__, prefix + "-obj.pickle") as fh:
            obj_data = pickle.load(fh)
        obj = cls(**obj_data)
        obj.make_models()
        with rs(__name__, prefix + "-training.h5") as fh:
            obj.training_model.load_weights(fh)
        with rs(__name__, prefix + "-infer-encoder.h5") as fh:
            obj.infer_encoder_model.load_weights(fh)
        with rs(__name__, prefix + "-infer-decoder.h5") as fh:
            obj.infer_decoder_model.load_weights(fh)
        return obj

    def save_model_weights(self, prefix):
        self.training_model.save_weights(prefix + "-training.h5")
        self.infer_encoder_model.save_weights(prefix + "-infer-encoder.h5")
        self.infer_decoder_model.save_weights(prefix + "-infer-decoder.h5")

    def load_model_weights(self, prefix):
        self.training_model.load_weights(prefix + "-training.h5")
        self.infer_encoder_model.load_weights(prefix + "-infer-encoder.h5")
        self.infer_decoder_model.load_weights(prefix + "-infer-decoder.h5")

    def fit(self, *args, **kwargs):
        callbacks = []
        if 'callbacks' in kwargs:
            kwargs['callbacks'].extend(callbacks)
        else:
            kwargs['callbacks'] = callbacks
        history = self.training_model.fit(*args, **kwargs)
        self.training_history = history.history
        return history

    def infer(self, input_seq):
        "Infers initial state for decoder from sequence"
        return self.infer_encoder_model.predict(input_seq)

    def decode_sequence(self, state_value, target_seq, transform=lambda x: x,
                        maxlen=20):
        """Decodes a sequence given an initial state.

        Yields (1, 1, vocab_len) array with next predicted item from sequence.

        Parameters
        ----------
        state_value : numpy.array
            Initial state value for decoder
        target_seq : numpy.array
            Initial target sequence
        transform : function
            Function to transform sample from decoder before yielding and
            feeding back into the decoder. A good place to do your np.argmax
            or temperature or whatever.
        maxlen : int
            Maximum desired length of decoded sequence
        """

        yield transform(target_seq)
        for i in range(maxlen):
            output, h = self.infer_decoder_model.predict(
                [target_seq, state_value])
            sampled = transform(output)
            yield sampled
            target_seq = sampled
            state_value = h

    def vectorize_src(self, t, maxlen):
        """Vectorize list of tokens from source vocabulary.

        Returns (1, maxlen, src_vocab_len) array vectorizing t from source
        vocab.
        """

        seq = np.zeros((1, maxlen, len(self.src_vocab)))
        for i, item in enumerate(t):
            for j in item:
                seq[0, i, self.src_vocab_idx_map[j]] = 1.
        return seq

    def vectorize_target(self, t, maxlen):
        """Vectorize list of tokens from target vocabulary.

        Returns (1, maxlen, target_vocab_len) array vectorizing t from target
        vocab.
        """

        seq = np.zeros((1, maxlen, len(self.target_vocab)))
        for i, item in enumerate(t):
            for j in item:
                seq[0, i, self.target_vocab_idx_map[j]] = 1.
        return seq

    def to_src_vocab(self, arr, maxlen):
        """
        Converts (1, maxlen, src_vocab_len) array to list of lists of src_vocab
        items where the corresponding index is greater than zero.
        """
        out = []
        for timestep in arr[0]:
            out.append([self.src_vocab[i] for i in np.where(timestep > 0)[0]])
        return out

    def to_target_vocab(self, arr, maxlen):
        """
        Converts (1, maxlen, src_vocab_len) array to list of lists of src_vocab
        items where the corresponding index is greater than zero.
        """
        out = []
        for timestep in arr[0]:
            out.append(
                    [self.target_vocab[i] for i in np.where(timestep > 0)[0]])
        return out


def softmax_maximum(t):
    zeros = np.zeros(t.shape)
    argmax = np.argmax(t[0, -1, :])
    zeros[0, 0, argmax] = 1.
    return zeros


def softmax_temperature(t, temp=0.5):
    zeros = np.zeros(t.shape)
    dist = t[0, -1, :].astype(np.float64) + 0.0001
    pred = np.log(dist) / temp
    exp_pred = np.exp(pred)
    pred = exp_pred / (np.sum(exp_pred))
    yhat = np.argmax(np.random.multinomial(1, pred, 1))
    zeros[0, 0, yhat] = 1.
    return zeros


def sigmoid_top_n(t, n=4, minval=0.0, threshold=True):
    """Get top n values from array (for output of sigmoid layer)

    Returns t with all but the n highest values in the array zeroed out.
    Only values higher than minval (default 0) will be included (potentially
    resulting in fewer than n items being included in the result). If
    threshold is included, all nonzero values (after sorting and check
    against minval) will be converted to 1.

    """
    zeros = np.zeros(t.shape)
    top_n = np.argsort(t[0, -1, :])[::-1][:n]
    zeros[0, 0, top_n] = t[0, 0, top_n]
    if threshold:
        return np.where(zeros > minval, 1., 0.)
    else:
        return np.where(zeros > minval, t, 0.)


def sigmoid_sample(t, squish=1., threshold=True):
    """Sample from multi-label probabilities in t.

    The 'squish' parameter pushes the values in t toward extremes:
    Values > 1 emphasize high/low probabilities, while values < 1
    move toward uniform distribution."""

    # guard against divide by zero
    t = np.where(t == 0, 0.0001, t)
    t = np.where(t == 1, 0.9999, t)
    probs = 1 / (1 + (np.exp(np.log((1/t)-1)) + np.exp(squish)))
    if threshold:
        return np.where(t > np.random.uniform(size=probs.shape), 1, 0)
    else:
        return np.where(t > np.random.uniform(size=probs.shape), t, 0)


pincelate/cmudictdata.py
from sklearn.model_selection import train_test_split
import numpy as np
import pincelate.featurephone as featurephone
import pronouncing as pr
import re
import itertools

# pronouncing loads data on demand, so let's demand it
pr.init_cmu()


def phoneme_feature_vocab():
    feat_vals = set(itertools.chain(*featurephone.phone_feature_map.values()))
    all_vals = list(feat_vals) + ["str"]  # "str" used to mark stressed vowels
    feature_vocab = sorted(all_vals)
    feature_vocab_idx_map = {k: i for i, k in enumerate(feature_vocab)}
    return feature_vocab, feature_vocab_idx_map


def max_phoneme_feature_len():
    # +2 to account for beginning/ending tokens
    return max((len(p.split()) for w, p in pr.pronunciations)) + 2


def orthography_vocab():
    letters = set(itertools.chain(*[list(a) for a, b in pr.pronunciations]))
    # ensure vocab item 0 is end of string
    letter_vocab = ["$", "^"] + list(sorted(letters))
    letter_vocab_idx_map = {k: i for i, k in enumerate(letter_vocab)}
    return letter_vocab, letter_vocab_idx_map


def max_orthography_len():
    # +2 to account for beginning/ending tokens
    return max((len(w) for w, p in pr.pronunciations)) + 2


def load():
    pr.init_cmu()  # pronouncing loads data on demand, so let's demand it

    feature_vocab, feature_vocab_idx_map = phoneme_feature_vocab()
    letter_vocab, letter_vocab_idx_map = orthography_vocab()
    max_phone_len = max_phoneme_feature_len()
    max_letter_len = max_orthography_len()

    # TODO: held-out data here

    letter_data = np.zeros(
        (len(pr.pronunciations), max_letter_len, len(letter_vocab)),
        dtype=np.float32
    )
    phonfeat_data = np.zeros(
        (len(pr.pronunciations), max_phone_len, len(feature_vocab)),
        dtype=np.float32
    )
    letter_target_data = np.zeros(
        (len(pr.pronunciations), max_letter_len, len(letter_vocab)),
        dtype=np.float32
    )
    phonfeat_target_data = np.zeros(
        (len(pr.pronunciations), max_phone_len, len(feature_vocab)),
        dtype=np.float32
    )

    for i, (word, phones) in enumerate(pr.pronunciations):

        # orthography: one-hot for each character index
        word = "^" + word + "$"
        for t, char in enumerate(word):
            letter_data[i, t, letter_vocab_idx_map[char]] = 1.
            if t > 0:
                letter_target_data[i, t-1, letter_vocab_idx_map[char]] = 1.

        # clean errant comments
        phones = re.sub(' #.*$', '', phones)

        # phonemes: multi-label k-hot phonetic features
        for t, phone in enumerate(["^"] + phones.split() + ["$"]):
            for ft in featurephone.phone_feature_map[phone.strip('012')] + \
                    (('str',) if re.search(r'[12]$', phone) else tuple()):
                phonfeat_data[i, t, feature_vocab_idx_map[ft]] = 1.
                if t > 0:
                    phonfeat_target_data[i, t-1, feature_vocab_idx_map[ft]] = 1

    return {
        'phoneme_feature_vocab': feature_vocab,
        'phoneme_feature_idx_map': feature_vocab_idx_map,
        'orthography_vocab': letter_vocab,
        'orthography_idx_map': letter_vocab_idx_map,
        'phoneme_feature_data': phonfeat_data,
        'phoneme_feature_target_data': phonfeat_target_data,
        'orthography_data': letter_data,
        'orthography_target_data': letter_target_data
    }


def tts(data, test_size=0.2, random_state=None):
    tts = train_test_split(data['orthography_data'],
                           data['orthography_target_data'],
                           data['phoneme_feature_data'],
                           data['phoneme_feature_target_data'],
                           test_size=test_size,
                           random_state=random_state)
    return {
        "orth_data_train": tts[0],
        "orth_data_valid": tts[1],
        "orth_target_train": tts[2],
        "orth_target_valid": tts[3],
        "phon_data_train": tts[4],
        "phon_data_valid": tts[5],
        "phon_target_train": tts[6],
        "phon_target_valid": tts[7]
    }


if __name__ == '__main__':
    import sys
    import pickle
    obj = load()
    pickle.dump(obj, sys.stdout.buffer)


pincelate/featurephone.py

phone_feature_map = {
    'M': ('blb', 'nas'),
    'P': ('vls', 'blb', 'stp'),
    'B': ('vcd', 'blb', 'stp'),
    'F': ('vls', 'lbd', 'frc'),
    'V': ('vcd', 'lbd', 'frc'),
    'TH': ('vls', 'dnt', 'frc'),
    'DH': ('vcd', 'dnt', 'frc'),
    'N': ('alv', 'nas'),
    'T': ('vls', 'alv', 'stp'),
    'D': ('vcd', 'alv', 'stp'),
    'S': ('vls', 'alv', 'frc'),
    'Z': ('vcd', 'alv', 'frc'),
    'R': ('alv', 'apr'),
    'L': ('alv', 'lat'),
    'SH': ('vls', 'pla', 'frc'),
    'ZH': ('vcd', 'pla', 'frc'),
    'Y': ('pal', 'apr'),
    'NG': ('vel', 'nas'),
    'K': ('vls', 'vel', 'stp'),
    'G': ('vcd', 'vel', 'stp'),
    'W': ('lbv', 'apr'),
    'HH': ('glt', 'apr'),
    'CH': ('vls', 'alv', 'stp', 'frc'),
    'JH': ('vcd', 'alv', 'stp', 'frc'),
    'AO': ('lmd', 'bck', 'rnd', 'vwl'),
    'AA': ('low', 'bck', 'unr', 'vwl'),
    'IY': ('hgh', 'fnt', 'unr', 'vwl'),
    'UW': ('hgh', 'bck', 'rnd', 'vwl'),
    'EH': ('lmd', 'fnt', 'unr', 'vwl'),
    'IH': ('smh', 'fnt', 'unr', 'vwl'),
    'UH': ('smh', 'bck', 'rnd', 'vwl'),
    'AH': ('mid', 'cnt', 'unr', 'vwl'),
    'AE': ('low', 'fnt', 'unr', 'vwl'),
    'EY': ('lmd', 'smh', 'fnt', 'unr', 'vwl'),
    'AY': ('low', 'smh', 'fnt', 'cnt', 'unr', 'vwl'),
    'OW': ('umd', 'smh', 'bck', 'rnd', 'vwl'),
    'AW': ('low', 'smh', 'bck', 'cnt', 'unr', 'rnd', 'vwl'),
    'OY': ('lmd', 'smh', 'bck', 'fnt', 'rnd', 'unr', 'vwl'),
    'ER': ('umd', 'cnt', 'rzd', 'vwl'),
    '^': ('beg',),
    '$': ('end',)
}

if __name__ == '__main__':
    import doctest
    doctest.testmod()


pincelate/train.py
# specialized script for training cmudict data.


def main(args):
    from pincelate.seq2seq import Seq2Seq
    import pincelate.cmudictdata

    print("loading data...")
    data = pincelate.cmudictdata.load()
    tts = pincelate.cmudictdata.tts(data, random_state=args.random_state)

    if args.src == 'orth':
        src_vocab = data['orthography_vocab']
        encoder_data_train = tts['orth_data_train']
        encoder_data_valid = tts['orth_data_valid']
    elif args.src == 'phon':
        src_vocab = data['phoneme_feature_vocab']
        encoder_data_train = tts['phon_data_train']
        encoder_data_valid = tts['phon_data_valid']

    if args.target == 'orth':
        target_vocab = data['orthography_vocab']
        decoder_data_train = tts['orth_data_train']
        decoder_data_valid = tts['orth_data_valid']
        target_data_train = tts['orth_target_train']
        target_data_valid = tts['orth_target_valid']
        activation = 'softmax'
        loss = 'categorical_crossentropy'
    elif args.target == 'phon':
        target_vocab = data['phoneme_feature_vocab']
        decoder_data_train = tts['phon_data_train']
        decoder_data_valid = tts['phon_data_valid']
        target_data_train = tts['phon_target_train']
        target_data_valid = tts['phon_target_valid']
        activation = 'sigmoid'
        loss = 'binary_crossentropy'

    params = {}
    for param in ['enc_rnn_units', 'dec_rnn_units', 'enc_rnn_dropout',
                  'dec_rnn_dropout', 'lr', 'decay', 'clipvalue', 'optimizer']:
        params[param] = getattr(args, param)

    params['bidi'] = not(args.unidirectional)
    params['vae'] = False  # reserved for future use

    if args.verbose:
        print(params)

    model = Seq2Seq(
            src_vocab=src_vocab,
            target_vocab=target_vocab,
            activation=activation,
            loss=loss,
            hps=params)

    model.make_models()
    model.compile_models()
    model.training_model.summary()
    history = model.fit(
            [encoder_data_train, decoder_data_train],
            target_data_train,
            batch_size=args.batch_size,
            epochs=args.epochs,
            validation_data=(
                [encoder_data_valid, decoder_data_valid],
                target_data_valid),
            verbose=1 if args.verbose else 2)
    print(history.history)
    print("saving model...")
    model.save(args.model_prefix)
    print("done")


if __name__ == '__main__':
    import argparse
    parser = argparse.ArgumentParser(
        description='Training script for Pincelate')

    parser.add_argument(
            '--model-prefix',
            type=str,
            help='prefix for saved models (directories must already exist!)',
            required=True)
    parser.add_argument(
            '--verbose',
            action='store_true',
            help='show keras progress bars (default to one line per epoch)')
    parser.add_argument(
            '--random-state',
            type=int,
            default=None,
            help='random state for train/test split')
    parser.add_argument(
            '--epochs',
            type=int,
            default=50,
            help='number of epochs to train')
    parser.add_argument(
            '--batch-size',
            type=int,
            default=128,
            help='batch size')
    parser.add_argument(
            '--src',
            choices=['orth', 'phon'],
            required=True,
            help='source sequences')
    parser.add_argument(
            '--target',
            choices=['orth', 'phon'],
            required=True,
            help='target sequences')
    parser.add_argument(
            '--unidirectional',
            action='store_true',
            help='unidirectional rnn (default is bidirectional)')
    parser.add_argument(
            '--enc-rnn-units',
            type=int,
            default=128,
            help='units in encoder RNN')
    parser.add_argument(
            '--dec-rnn-units',
            type=int,
            default=128,
            help='units in decoder RNN')
    parser.add_argument(
            '--enc-rnn-dropout',
            type=float,
            default=0.2,
            help='recurrent dropout in encoder RNN')
    parser.add_argument(
            '--dec-rnn-dropout',
            type=float,
            default=0.2,
            help='recurrent dropout in decoder RNN')
    parser.add_argument(
            '--optimizer',
            type=str,
            choices=['adam', 'rmsprop'],
            default='adam',
            help='optimizer (rmsprop or adam)')
    parser.add_argument(
            '--lr',
            type=float,
            default=0.001,
            help='learning rate for optimizer')
    parser.add_argument(
            '--decay',
            type=float,
            default=1e-6,
            help='learning rate decay for optimizer')
    parser.add_argument(
            '--clipvalue',
            type=float,
            default=1.0,
            help='clip value for optimizer')

    args = parser.parse_args()
    main(args)
